Add support to serialize qtensor in JIT.

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23356 Add support to serialize qtensor in JIT.**

Adds qtensor specific fields to the proto file so that they get serialized into the model.json

Differential Revision: [D16473237](https://our.internmc.facebook.com/intern/diff/D16473237/)

supriyar committed:
Add support to serialize qtensor in JIT.

Adds qtensor specific fields to the proto file so that they get serialized into the model.json

Differential Revision: [D16473237](https://our.internmc.facebook.com/intern/diff/D16473237/)

dzhulgakov commented on torch/csrc/jit/import.cpp:
```diff
@@ -258,15 +258,25 @@ at::Tensor ScriptModuleDeserializer::loadTensor(
-        at::empty(
-            {0}, c10::TensorOptions(type).device(storage_it->second.device()))
-            .set_(storage_it->second, tensor_proto.offset(), dims, strides);
+  if (tensor_proto.is_quantized()) {
```
@jerryzh - do we support on-device quantized tensors? if so  - we should merge it with the if below

jerryzh168 replied:
what is on-device quantized tensor?

dzhulgakov replied:
Sorry, I meant e.g. CUDA quantized tensor. But given that this code is likely going to change in the future with serialization overhaul, I guess it's fine to have it as is.

dzhulgakov commented on torch/csrc/jit/import.cpp:
```diff
@@ -258,15 +258,25 @@ at::Tensor ScriptModuleDeserializer::loadTensor(
+  else {
+    if (device.type() == at::DeviceType::CPU) {
+      result =
+          at::empty({0}, at::CPU(type).options())
```
nit: it can probably be refactored by pulling out TensorOptions to a separate line and then just having a single call to at::empty (or at::_empty_affine_quantized)

supriyar replied:
I moved the is_quantized check inside CPU device type check.

supriyar committed:
Update on "Add support to serialize qtensor in JIT."


Adds qtensor specific fields to the proto file so that they get serialized into the model.json

Differential Revision: [D16473237](https://our.internmc.facebook.com/intern/diff/D16473237/)

supriyar commented:
@jerryzh168 please review

jerryzh168 reviewed (approved):
LGTM

supriyar committed:
Update on "Add support to serialize qtensor in JIT."


Adds qtensor specific fields to the proto file so that they get serialized into the model.json

Differential Revision: [D16473237](https://our.internmc.facebook.com/intern/diff/D16473237/)

jerryzh168 reviewed (approved):
looks pretty good, feel free to land if sandcastle and ossci passes

jerryzh168 commented:
actually, should you change the TypeMetaToDataType function to map quantized types to non-quantized types?

supriyar committed:
Update on "Add support to serialize qtensor in JIT."


Adds qtensor specific fields to the proto file so that they get serialized into the model.json

Differential Revision: [D16473237](https://our.internmc.facebook.com/intern/diff/D16473237/)
